File: packages/dsh-sag/runtime/src/dsh_sag_runtime/protocol.py
"""Versioned JSON-RPC 2.0 wire messages used over stdio."""

from __future__ import annotations

import json
from typing import Any, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, field_validator

RPC_PROTOCOL_VERSION = "1.0"
REQUIRED_ENGINE_VERSION = "0.10.0"
MAX_FRAME_BYTES = 8 * 1024 * 1024


class WireModel(BaseModel):
    """Immutable strict wire model with camel-case serialization aliases."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)


class InitializeParams(WireModel):
    protocol_version: Literal["1.0"] = Field(alias="protocolVersion")


class SearchParams(WireModel):
    query: str = Field(min_length=1, max_length=32_000)
    namespaces: tuple[str, ...] = Field(min_length=1)
    mode: Literal["fast", "precise"]
    limit: int = Field(ge=1, le=50)

    @field_validator("namespaces")
    @classmethod
    def validate_namespaces(cls, values: tuple[str, ...]) -> tuple[str, ...]:
        if any(not value or len(value) > 36 for value in values):
            raise ValueError("namespace ids must contain 1 to 36 characters")
        return values


class ReadParams(WireModel):
    evidence_ref: str = Field(alias="evidenceRef", min_length=1, max_length=4096)
    offset: int = Field(ge=0)
    max_chars: int = Field(alias="maxChars", ge=1, le=200_000)
    include_events: bool = Field(alias="includeEvents")


class CancelParams(WireModel):
    id: int = Field(ge=0)


class EmptyParams(WireModel):
    pass


class InitializeRequest(WireModel):
    jsonrpc: Literal["2.0"]
    id: int = Field(ge=0)
    method: Literal["initialize"]
    params: InitializeParams


class SearchRequest(WireModel):
    jsonrpc: Literal["2.0"]
    id: int = Field(ge=0)
    method: Literal["search"]
    params: SearchParams


class ReadRequest(WireModel):
    jsonrpc: Literal["2.0"]
    id: int = Field(ge=0)
    method: Literal["read"]
    params: ReadParams


class ShutdownRequest(WireModel):
    jsonrpc: Literal["2.0"]
    id: int = Field(ge=0)
    method: Literal["shutdown"]
    params: EmptyParams


class CancelRequest(WireModel):
    jsonrpc: Literal["2.0"]
    method: Literal["$/cancelRequest"]
    params: CancelParams


RpcRequest: TypeAlias = InitializeRequest | SearchRequest | ReadRequest | ShutdownRequest | CancelRequest

_REQUEST_MODELS: dict[str, type[WireModel]] = {
    "initialize": InitializeRequest,
    "search": SearchRequest,
    "read": ReadRequest,
    "shutdown": ShutdownRequest,
    "$/cancelRequest": CancelRequest,
}


class RpcErrorData(WireModel):
    code: str
    operation: str | None = None
    stage: str | None = None
    retryable: bool | None = None
    provider: str | None = None
    item_id: str | None = Field(default=None, alias="itemId")
    message: str | None = None
    details: dict[str, Any] | None = None


class RpcError(WireModel):
    code: int
    message: str
    data: RpcErrorData | None = None


class RpcSuccessResponse(WireModel):
    jsonrpc: Literal["2.0"] = "2.0"
    id: int = Field(ge=0)
    result: dict[str, Any]


class RpcErrorResponse(WireModel):
    jsonrpc: Literal["2.0"] = "2.0"
    id: int = Field(ge=0)
    error: RpcError


RpcResponse: TypeAlias = RpcSuccessResponse | RpcErrorResponse


def decode_request(frame: bytes) -> RpcRequest:
    """Decode one bounded request frame and reject unknown methods or fields."""
    if len(frame) > MAX_FRAME_BYTES:
        raise ValueError("request frame exceeds 8 MiB")
    raw = json.loads(frame)
    if not isinstance(raw, dict):
        raise ValueError("request must be an object")
    method = raw.get("method")
    model = _REQUEST_MODELS.get(method) if isinstance(method, str) else None
    if model is None:
        raise ValueError("unsupported JSON-RPC method")
    return model.model_validate(raw)  # type: ignore[return-value]


def success_response(request_id: int, result: dict[str, Any]) -> RpcSuccessResponse:
    """Create a successful correlated response."""
    return RpcSuccessResponse(id=request_id, result=result)


def error_response(request_id: int, code: int, message: str, data: RpcErrorData) -> RpcErrorResponse:
    """Create a failed correlated response with safe structured data."""
    return RpcErrorResponse(id=request_id, error=RpcError(code=code, message=message, data=data))


def encode_response(response: RpcResponse) -> bytes:
    """Serialize one response as a compact NDJSON frame."""
    payload = response.model_dump_json(by_alias=True, exclude_none=True)
    frame = f"{payload}\n".encode()
    if len(frame) > MAX_FRAME_BYTES:
        raise ValueError("response frame exceeds 8 MiB")
    return frame

File: packages/dsh-sag/src/runtime/protocol.ts
import { Buffer } from 'node:buffer'
import type { SagEvidenceRef, SagNamespaceId } from '../brand.ts'

export const RPC_PROTOCOL_VERSION = '1.0' as const
export const REQUIRED_ENGINE_VERSION = '0.10.0' as const
export const MAX_FRAME_BYTES = 8 * 1024 * 1024

export type RpcId = number
export type SearchMode = 'fast' | 'precise'

export interface InitializeResult {
  readonly protocolVersion: string
  readonly engineVersion: string
  readonly health: 'available' | 'degraded' | 'unavailable'
  readonly evidenceRead: boolean
  readonly namespaces: readonly string[]
}

export interface SearchEvidence {
  readonly evidenceRef: SagEvidenceRef
  readonly namespaceId: SagNamespaceId
  readonly sourceId: string
  readonly title: string
  readonly excerpt: string
  readonly score?: number
}

export interface SearchResult {
  readonly query: string
  readonly evidences: readonly SearchEvidence[]
}

export interface ReadEvent {
  readonly id?: string
  readonly title?: string
  readonly summary?: string
  readonly category?: string
  readonly rank?: number
}

export interface ReadResult {
  readonly title: string
  readonly content: string
  readonly offset: number
  readonly nextOffset?: number
  readonly totalChars: number
  readonly events?: readonly ReadEvent[]
}

export interface RpcErrorData {
  readonly code: string
  readonly operation?: string
  readonly stage?: string
  readonly retryable?: boolean
  readonly provider?: string
  readonly itemId?: string
  readonly message?: string
  readonly details?: Readonly<Record<string, unknown>>
}

export interface RpcError {
  readonly code: number
  readonly message: string
  readonly data?: RpcErrorData
}

export type RpcResult = InitializeResult | SearchResult | ReadResult | Readonly<Record<string, never>>
export type RpcResponse =
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly result: RpcResult }
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly error: RpcError }

export type RpcRequest =
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly method: 'initialize'; readonly params: { readonly protocolVersion: '1.0' } }
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly method: 'search'; readonly params: { readonly query: string; readonly namespaces: readonly string[]; readonly mode: SearchMode; readonly limit: number } }
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly method: 'read'; readonly params: { readonly evidenceRef: string; readonly offset: number; readonly maxChars: number; readonly includeEvents: boolean } }
  | { readonly jsonrpc: '2.0'; readonly id: RpcId; readonly method: 'shutdown'; readonly params: Readonly<Record<string, never>> }
  | { readonly jsonrpc: '2.0'; readonly method: '$/cancelRequest'; readonly params: { readonly id: RpcId } }

/** A safe protocol failure that identifies only the frame correlation point. */
export class SagProtocolError extends Error {
  override readonly name = 'SagProtocolError'
}

function object(value: unknown, label: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new SagProtocolError(`${label} must be an object`)
  }
  return value as Record<string, unknown>
}

function exactKeys(value: Record<string, unknown>, required: readonly string[], optional: readonly string[], label: string): void {
  const allowed = new Set([...required, ...optional])
  if (required.some(key => !(key in value)) || Object.keys(value).some(key => !allowed.has(key))) {
    throw new SagProtocolError(`${label} contains invalid fields`)
  }
}

function validId(value: unknown): value is RpcId {
  return typeof value === 'number' && Number.isSafeInteger(value) && value >= 0
}

function validateResult(value: unknown, id: number): RpcResult {
  const result = object(value, `response id ${id} result`)
  const keys = Object.keys(result)
  if (keys.length === 0) return {}
  if ('protocolVersion' in result) {
    exactKeys(result, ['protocolVersion', 'engineVersion', 'health', 'evidenceRead', 'namespaces'], [], `response id ${id} initialize result`)
    if (typeof result.protocolVersion !== 'string' || typeof result.engineVersion !== 'string'
      || !['available', 'degraded', 'unavailable'].includes(String(result.health))
      || typeof result.evidenceRead !== 'boolean'
      || !Array.isArray(result.namespaces) || !result.namespaces.every(item => typeof item === 'string')) {
      throw new SagProtocolError(`response id ${id} initialize result is invalid`)
    }
    return result as unknown as InitializeResult
  }
  if ('evidences' in result) {
    exactKeys(result, ['query', 'evidences'], [], `response id ${id} search result`)
    if (typeof result.query !== 'string' || !Array.isArray(result.evidences)) {
      throw new SagProtocolError(`response id ${id} search result is invalid`)
    }
    for (const raw of result.evidences) {
      const evidence = object(raw, `response id ${id} evidence`)
      exactKeys(evidence, ['evidenceRef', 'namespaceId', 'sourceId', 'title', 'excerpt'], ['score'], `response id ${id} evidence`)
      if (['evidenceRef', 'namespaceId', 'sourceId', 'title', 'excerpt'].some(key => typeof evidence[key] !== 'string')
        || ('score' in evidence && typeof evidence.score !== 'number')) {
        throw new SagProtocolError(`response id ${id} evidence is invalid`)
      }
    }
    return result as unknown as SearchResult
  }
  if ('content' in result) {
    exactKeys(result, ['title', 'content', 'offset', 'totalChars'], ['nextOffset', 'events'], `response id ${id} read result`)
    if (typeof result.title !== 'string' || typeof result.content !== 'string'
      || !Number.isSafeInteger(result.offset) || (result.offset as number) < 0
      || !Number.isSafeInteger(result.totalChars) || (result.totalChars as number) < 0
      || ('nextOffset' in result && (!Number.isSafeInteger(result.nextOffset) || (result.nextOffset as number) < 0))
      || ('events' in result && !Array.isArray(result.events))) {
      throw new SagProtocolError(`response id ${id} read result is invalid`)
    }
    return result as unknown as ReadResult
  }
  throw new SagProtocolError(`response id ${id} result contains invalid fields`)
}

/** Decode and validate one JSON-RPC response frame from the Python process. */
export function decodeResponse(line: string): RpcResponse {
  if (Buffer.byteLength(line, 'utf8') > MAX_FRAME_BYTES) {
    throw new SagProtocolError('response frame exceeds 8 MiB')
  }
  let raw: unknown
  try {
    raw = JSON.parse(line)
  } catch {
    throw new SagProtocolError('response frame is not valid JSON')
  }
  const response = object(raw, 'response')
  const idLabel = validId(response.id) ? ` id ${response.id}` : ''
  exactKeys(response, ['jsonrpc', 'id'], ['result', 'error'], `response${idLabel}`)
  if (response.jsonrpc !== '2.0' || !validId(response.id)) {
    throw new SagProtocolError(`response${idLabel} has invalid JSON-RPC metadata`)
  }
  if (('result' in response) === ('error' in response)) {
    throw new SagProtocolError(`response id ${response.id} must contain exactly one of result or error`)
  }
  if ('result' in response) {
    return { jsonrpc: '2.0', id: response.id, result: validateResult(response.result, response.id) }
  }
  const error = object(response.error, `response id ${response.id} error`)
  exactKeys(error, ['code', 'message'], ['data'], `response id ${response.id} error`)
  if (!Number.isInteger(error.code) || typeof error.message !== 'string') {
    throw new SagProtocolError(`response id ${response.id} error is invalid`)
  }
  let data: RpcErrorData | undefined
  if ('data' in error) {
    const rawData = object(error.data, `response id ${response.id} error data`)
    exactKeys(rawData, ['code'], ['operation', 'stage', 'retryable', 'provider', 'itemId', 'message', 'details'], `response id ${response.id} error data`)
    if (typeof rawData.code !== 'string') throw new SagProtocolError(`response id ${response.id} error data is invalid`)
    data = rawData as unknown as RpcErrorData
  }
  return {
    jsonrpc: '2.0',
    id: response.id,
    error: { code: error.code as number, message: error.message, ...(data ? { data } : {}) },
  }
}

/** Encode one validated request as a single NDJSON frame. */
export function encodeRequest(request: RpcRequest): string {
  const line = `${JSON.stringify(request)}\n`
  if (Buffer.byteLength(line, 'utf8') > MAX_FRAME_BYTES) {
    throw new SagProtocolError('request frame exceeds 8 MiB')
  }
  return line
}
